pkg/azuredx/azureauth/auth.go

```diff
@@ -0,0 +1,152 @@
+// Package azureauth provides authorization utilities for the Azure cloud.
+package azureauth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/models"
+)
+
+var onBehalfOfLatencySeconds = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "grafana",
+	Subsystem: "plugin_adx",
+	Name:      "obo_latency_seconds",
+	Help:      "On-behalf-of token exchange duration.",
+	Buckets:   []float64{0.01, 0.05, 0.10, 0.20, 0.40, 1.00},
+}, []string{"http_status"})
+
+func init() {
+	prometheus.MustRegister(onBehalfOfLatencySeconds)
+}
+
+// ServiceCredentials provides authorization for cloud service usage.
+type ServiceCredentials struct {
+	models.DatasourceSettings
+	// PostForm is an http.Client method.
+	PostForm func(url string, data url.Values) (*http.Response, error)
+	// ServicePrincipalToken resolves service credentials.
+	ServicePrincipalToken func(context.Context) (token string, err error)
+}
+
+// ServicePrincipalAuthorization returns an HTTP Authorization-header value which
+// represents the service principal.
+func (c *ServiceCredentials) ServicePrincipalAuthorization(ctx context.Context) (string, error) {
+	token, err := c.ServicePrincipalToken(ctx)
+	if err != nil {
+		return "", fmt.Errorf("service principal token unavailable: %w", err)
+	}
+	return "Bearer " + token, nil
+}
+
+// QueryDataAuthorization returns an HTTP Authorization-header value which
+// represents the respective request.
+func (c *ServiceCredentials) QueryDataAuthorization(req *backend.QueryDataRequest) (string, error) {
+	switch {
+	case c.OnBehalfOf:
+		return c.queryDataOnBehalfOf(req)
+
+	// TODO(pascaldekloe): case c.ClientCreds [oauthPassThru]
+
+	default:
+		backend.Logger.Debug("using service principal token for data request")
+		ctx := context.Background()
```

Comment:
QueryDataAuthorization should accept context as an argument and use that instead.